### PR TITLE
Test that text input is entered after child window is closed

### PR DESCRIPTION
**Pull request by wmww**
This is supposed to test https://github.com/MirServer/mir/issues/2189, but it passes for some reason. A problem for next week.

### CODE REVIEW

**Comment by wmww** [comment]
Ah! I forgot to actually set the parent on the child. Fixed. (and by fixed, I mean the test is now failing as it should)

**Comment by wmww** [comment]
https://github.com/MirServer/mir/pull/2223 fixes this test on Mir.

**Review by RAOF — CHANGES_REQUESTED** [review]
Oops! Looks like I left a review pending :grimacing: 

One non-blocking stylistic thing wrt `operator wl_surface*()`, one question about what we're testing that might require changes.

File: tests/text_input_v3_with_input_method_v2.cpp [hunk 1]
```diff
@@ -173,3 +174,83 @@ TEST_F(TextInputV3WithInputMethodV2Test, input_method_can_send_preedit)
     input_client.roundtrip();
     app_client.roundtrip();
 }
+
+TEST_F(TextInputV3WithInputMethodV2Test, text_input_does_not_enter_non_grabbing_popup)
+{
+    auto parent_surface = std::make_unique<wlcs::Surface>(app_client);
+    EXPECT_CALL(text_input, enter(parent_surface->operator wl_surface*()));
```
[the remark this answers is not in the document]
Comment (wmww) [reply tests/text_input_v3_with_input_method_v2.cpp:181]: I do believe some sort of explicit conversion is necessary. I, personally, like the explicit call to operator. Is there any reason to prefer static cast?

File: tests/text_input_v3_with_input_method_v2.cpp [hunk 2]
```diff
@@ -173,3 +174,83 @@ TEST_F(TextInputV3WithInputMethodV2Test, input_method_can_send_preedit)
...
+    auto parent_xdg_surface = std::make_shared<wlcs::XdgSurfaceStable>(app_client, *parent_surface);
+    auto parent_xdg_toplevel = std::make_shared<wlcs::XdgToplevelStable>(*parent_xdg_surface);
+    parent_surface->attach_visible_buffer(20, 20);
+    app_client.roundtrip();
+    Mock::VerifyAndClearExpectations(&text_input);
+    auto child_surface = std::make_unique<wlcs::Surface>(app_client);
+    EXPECT_CALL(text_input, leave(parent_surface->operator wl_surface*())).Times(0);
+    EXPECT_CALL(text_input, enter(child_surface->operator wl_surface*())).Times(0);
```
[the remark this answers is not in the document]
Comment (wmww) [reply tests/text_input_v3_with_input_method_v2.cpp]: Done

File: tests/text_input_v3_with_input_method_v2.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (AlanGriffiths) [reply tests/text_input_v3_with_input_method_v2.cpp:181]: > I, personally, like the explicit call to operator.

I, personally, find it strange to invoke operators by calling the operator function by name. The point of such functions is that they can be invoked using operator syntax.

If the intended use is to call the function by name then there are better names.

I'd be tempted to introduce a helper function:

```c++
auto as_wl_surface(wlcs::Surface const& s) -> wl_surface*
{
    return s;
}
```
```suggestion
    EXPECT_CALL(text_input, enter(as_wl_surface((*parent_surface)));
```

[the remark this answers is not in the document]
Comment (wmww) [reply tests/text_input_v3_with_input_method_v2.cpp]: Is the latest update satisfactory?

**Comment by wmww** [comment]
Ready for review

File: tests/text_input_v3_with_input_method_v2.cpp [hunk 3]
```diff
@@ -213,3 +214,83 @@ TEST_F(TextInputV3WithInputMethodV2Test, input_method_can_send_preedit)
...
+
+    // This is needed to get a serial, which will be used later on
+    auto pointer = the_server().create_pointer();
+    pointer.move_to(2, 2);
+    pointer.left_click();
+    app_client.roundtrip();
+
+    auto child_surface = std::make_unique<wlcs::Surface>(app_client);
+    EXPECT_CALL(text_input, leave(parent_surface->wl_surface()));
+    EXPECT_CALL(text_input, enter(child_surface->wl_surface()));
```
Comment (RAOF) [inline tests/text_input_v3_with_input_method_v2.cpp:258]: Non-blocking, but this ordering is a protocol requirement, so we should probably enforce it.

File: tests/text_input_v3_with_input_method_v2.cpp [hunk 4]
```diff
@@ -213,3 +214,83 @@ TEST_F(TextInputV3WithInputMethodV2Test, input_method_can_send_preedit)
...
+    auto parent_xdg_surface = std::make_shared<wlcs::XdgSurfaceStable>(app_client, *parent_surface);
+    auto parent_xdg_toplevel = std::make_shared<wlcs::XdgToplevelStable>(*parent_xdg_surface);
+    parent_surface->attach_visible_buffer(20, 20);
+    app_client.roundtrip();
+    Mock::VerifyAndClearExpectations(&text_input);
+    auto child_surface = std::make_unique<wlcs::Surface>(app_client);
+    EXPECT_CALL(text_input, leave(_)).Times(0);
+    EXPECT_CALL(text_input, enter(_)).Times(0);
+    auto child_xdg_surface = std::make_shared<wlcs::XdgSurfaceStable>(app_client, *child_surface);
+    auto child_xdg_toplevel = std::make_shared<wlcs::XdgPopupStable>(
```
Comment (RAOF) [inline tests/text_input_v3_with_input_method_v2.cpp:231]: Nit: `child_xdg_toplevel` -> `child_xdg_popup`

**Review by RAOF — APPROVED** [review]
A couple of nits, but this can land.

Feel free to fix them up, or not.

(I *think* the workflow for this now is just to approve and then let Sophie hit the “merge” button?)

**Comment by Saviq** [comment]
> (I _think_ the workflow for this now is just to approve and then let Sophie hit the “merge” button?)

Not before #289 :)

**Pull request merged**